### ISSUE TITLE
Typesetting issues in sile-0.10.15.pdf

### ISSUE THREAD

**Issue by zavislak** [sile-typesetter/sile#1174]
I'm new to SILE so I can't debug, but I wanted to note the following oddities:

On page 4 the spacing between two paragraphs is insufficient, probably due to the placement of the figure:

![image](https://user-images.githubusercontent.com/371987/120570526-3338bb00-c3cd-11eb-8f9a-9bbb13ad0890.png)

On page 74 in the two column demo, the left column has a vertical offset that takes the text out of the frame:

![image](https://user-images.githubusercontent.com/371987/120570642-7004b200-c3cd-11eb-9fae-be0ba40e495e.png)

**Comment by alerque** [comment]
Thanks for taking the time to report these @zavislak! (And welcome to SILE. Don't be afraid to post about any trouble you run into here even if it's a question about how to accomplish X and not strictly a bug report).

The first issue here is a duplicate of #394. Drop-caps (floats) and everything to do with them are a long-standing disaster situation. Lets leave that out of this issue and comment on the old tracking issue if needed.

The second issue I don't remember seeing reported before. Lets make this issue about tracking down that problem.

**Comment by Omikhleia** [comment]
_"A one-year old unaddressed issues...", thinks Omikhleia. "Hey, it's time to **pinpoint** it!_" :rofl: 

Let's recap:
![image](https://user-images.githubusercontent.com/18075640/177049280-0f912354-1d61-454f-ba06-47c6a2481c15.png)

Problem (2), i.e. some small frame overflow at the bottom, is easy to understand - it's a mix of two thing (fixed baseline skip from very early in `sile.sil`, aggravated by #1404). "_I am not going to investigate it further._"

But problem (1), i.e. some extra vertical space in the columns, is much more interesting. 
If one looks closely, the issue is not only on the first column (where it is indeed very visible), but also on the second one. Erm. Wait! Looking even more closely, it's actually all the frames top which are affected, from a certain point in the document. Let's use our eyes:
![image](https://user-images.githubusercontent.com/18075640/177049921-5c1bcafe-2000-4a57-8156-6cec7d4cdd0f.png)

We just need to track where this weird thing started occurring...  _Omikhleia puts on his spectacles... "**Gotcha**!_". 

**And the culprit is**.... That good old **grid** package, via it's package documentation.

When grid typesetting is enabled (`\grid[...]`), the code overrides a few typesetter methods. When grid typesetting is disabled (`\no-grid`), it restores these typesetter methods. We should be safe and expect having returned to the previous state... But wait, there's one thing still active! That frame hook we registered: `SILE.typesetter:registerNewFrameHook(startGridInFrame)`... It is still doing it's things far after we left the grid...

A workaround is actually doable here:
```
local function startGridInFrame (typesetter)
  if not SILE.typesetter.state.grid then return end -- EEK! HACK! HOOK!
  local queue = typesetter.state.outputQueue
  ...
```
**Problem solved**. Er. Root cause identified, with a way to work around it, rather...

Still, there are other bits of code smell... E.g.  `SILE.defaultTypesetter:pushBack` actually tests this grid state variable. Where is our neat separation of concern, if the _standard_ default typesetter has a dependency on some intimate workings of a package, that also had its overridden methods?

Also, this may raise an interesting question for those interested in idempotent things (hi there, @alerque)... We have ways to register hooks, but not to unregister them...

**Comment by alerque** [comment]
Re-unregistering stuff. Yup the idempotent overhauls knew that was coming! Everything that is *done* should have an *undo*. The most obvious one being a bulk `:unloadPackage()` to go along with `:loadPackage()`. There are not *so many* documents in the wild where this becomes necessary, but while working on package management stuff for `v0.14.0` I've been thinking the most useful case is going to be when two or more 3rd party packages want to muck with the same internals but conflict with each-other. A case in point is the color-fonts package and font-fallback package. I hacked together some tradoffs so you can switch back and forth, but all the refactoring I'm doing has an eventual unloader in mind.

The manual is a pretty good stress test for this in core packages, but we're had to hack around it being problematic in lots of places. We *should* be able to load a package, demo it's use, then move on with everything back the way it was. We're not there yet but *much* closer than 2 releases ago.

**Comment by Omikhleia** [comment]
For the sake of backlog cleaning, I _think_ we could close this issue:
- The first case is already tracked as  #860
- The second case was fixed in #1501
- The last case is already tracked as #1404
